lib/BSON.pm

```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
     return $hash;
 }
 
+=method perl_to_extjson
+
+    use JSON::MaybeXS;
+    my $ext = $bson->perl_to_extjson($data);
+    my $json = encode_json($ext, \%options);
+
+Takes a perl data structure and turns it into an Extended JSON
+structure. Note that the structure will still have to be serialized.
+
+Possible options are:
+
+=for :list
+* C<relaxed> A boolean indicating wether relaxed extended JSON should
+be generated. If not set, the value is defaulted to the C<BSON_EXTJSON>
+environment variable.
+
+=cut
+
+sub perl_to_extjson {
+    my ($class, $data, $options) = @_;
+
+    local $ENV{BSON_EXTJSON} = !$options->{relaxed}
+        if exists $options->{relaxed};
+
+    if (not defined $data) {
+        return undef;
+    }
+
+    if (blessed($data) and $data->can('TO_JSON')) {
+        my $json_data = $data->TO_JSON;
+        return $json_data;
+    }
+
+    if (not ref $data) {
+
+        if (!$ENV{BSON_EXTJSON}) {
+            return $data;
+        }
+        else {
+            if (looks_like_number($data)) {
+                if ($data =~ m{\A-?[0-9_]+\z}) {
+                    if ($data <= $max_int32) {
+                        return { '$numberInt' => "$data" };
+                    }
+                    else {
+                        return { '$numberLong' => "$data" };
+                    }
+                }
+                else {
+                    return { '$numberDouble' => 'Infinity' }
+                        if $data eq 'Inf';
+                    return { '$numberDouble' => '-Infinity' }
+                        if $data eq '-Inf';
+                    return { '$numberDouble' => 'NaN' }
+                        if $data eq 'NaN';
+                    my $value = "$data";
+                    $value = $value / 1.0;
+                    return { '$numberDouble' => "$value" };
+                }
+            }
+
+            return $data;
+        }
+    }
+
+    if (boolean::isBoolean($data)) {
+        return $data;
+    }
+
+    if (ref $data eq 'HASH') {
+        if (exists $data->{'$type'}) {
```

Comment:
I'm not sure this is needed.  The special handling for `$type` is needed when parsing ExtJSON but shouldn't need any special work for encoding.  (Unless this is handling some recursive situation from encoding Bytes/Binary)